nlmod/observations.py

```diff
@@ -0,0 +1,232 @@
+import numpy as np
+import xarray as xr
+from scipy.spatial import Delaunay
+
+from nlmod.dims import grid
+
+
+def compute_interpolation_weights(xy, uv, d=2):
+    """Calculate interpolation weights for barycentric interpolation [1]_.
+
+    Parameters
+    ----------
+    xy : np.array
+        array containing x-coordinates in first column and y-coordinates
+        in second column
+    uv : np.array
+        array containing coordinates at which interpolation weights should
+        be calculated, x-coordinates in first column and y-coordinates in second column
+    d : int, optional
+        dimension of data (the default is 2, which works for 2D data)
+
+    Returns
+    -------
+    vertices: np.array
+        array containing interpolation vertices (len(xy), 3)
+    weights: np.array
+        array containing interpolation weights per point (len(xy), 3)
+
+
+    References
+    ----------
+    .. [1] https://stackoverflow.com/questions/20915502/
+    speedup-scipy-griddata-for-multiple-interpolations-between-two-irregular-grids
+    """
+    tri = Delaunay(xy)
+    simplex = tri.find_simplex(uv)
+    vertices = np.take(tri.simplices, simplex, axis=0)
+    temp = np.take(tri.transform, simplex, axis=0)
+    delta = uv - temp[:, d]
+    bary = np.einsum("njk,nk->nj", temp[:, :d, :], delta)
+    return vertices, np.hstack((bary, 1 - bary.sum(axis=1, keepdims=True)))
+
+
+def interpolate_vertices_weights(values, vertices, weights, fill_value=np.nan):
+    """Interpolate values at locations given computed vertices and weights.
+
+    As calculated by compute_barycentric_interpolation_weights function [2]_.
+
+    Parameters
+    ----------
+    values : np.array
+        array containing values to interpolate
+    vtx : np.array
+        array containing interpolation vertices, see
+        compute_barycentric_interpolation_weights()
+    wts : np.array
+        array containing interpolation weights, see
+        compute_barycentric_interpolation_weights()
+    fill_value : float
+        fill value for points that have to be extrapolated (e.g. at or
+        beyond edges of the known points)
+
+    Returns
+    -------
+    arr: np.array
+        array containing interpolated values at locations as given by
+        vtx and wts
+
+    References
+    ----------
+    .. [2] https://stackoverflow.com/questions/20915502/
+    speedup-scipy-griddata-for-multiple-interpolations-between-two-irregular-grids
+    """
+    ret = np.einsum("nj,nj->n", np.take(values, vertices), weights)
+    ret[np.any(weights < 0, axis=1)] = fill_value
+    return ret
+
+
+def interpolate_points(da, xi, yi, x="x", y="y"):
+    """Interpolate values at locations given by x and y coordinates.
```

Comment:
When to use this one and when to use `interpolate_points_ds`?

Reply:
This one works for a a simple array and x, y, xi and yi passed as arrays. I see this as a simple version thats easier to use for simple tasks, for example, if all your stuff is in numpy for example.